proposals/kapp-controller/004-health-status-reporting/README.md

```diff
@@ -0,0 +1,76 @@
+
+1. The controller reconciling the App CR needs to dynamically set up watches for the resources being deployed by the package. 
+2. Introduce a `Healthy` condition in App CR's `status` [field][app_cr_status].
+
+This field would be set to:
+1. `True`: if all the installed resources are healthy.
+2. `False`: if there is any error during the `fetch`, `install`, `deploy` step or if any of the resources are unhelathy.
+
+The proposed set to criteria to determine if a resource is healthy or not is described below:
+
+1. A Pod resource will be healthy if/when: 
+- `.status.Phase` is Running or Completed.
+- If `.status.Phase` is Running, "Available" status condition is True.
+
+2. A ReplicationController resource will be healthy if/when:
+- `.spec.replicas` matches `.status.availableReplicas`
+- `ReplicaFailure` type condition is not present in the status.
+
+3. A ReplicaSet resource will be healthy if/when:
+- `.spec.replicas` matches `.status.availableReplicas`.
+- `ReplicaFailure` type condition is not present in the status.
+
+3. A Deployment resource will be healthy if/when:
+- `.spec.replicas` matches `.status.availableReplicas`.
+- `Available` type condition is true.
+
+4. A StatefulSet resource will be healthy if/when:
+- `.spec.replicas` matches `.status.availableReplicas`.
+- A DaemonSet resource will be healthy if/when:
+`.status.desiredNumberScheduled` matches `.status.numberAvailable`
+
+5. An APIService resource will be healthy if/when: 
+- `Available` type condition in status is true.
+
+6. A CustomResourceDefinition resource will be healthy if/when:
+- `StoredVersions` has the expected API version for the CRD.
+
+7. All other unspecified resources will be considered healthy.
+
+If any of the watched resource is unhealthy, the `Message` field of the healthy condition will have the statuses of the unhealthy resources ordered lexicographically. 
+
+Since the resources deployed by the App reconciler have informers created for them, any change in the resource state will trigger a reconcile that in turn will re-evaluate the health of all resources. 
+
+#### Use Case: Monitoring the state of resources
+
+Kapp currently has the `inspect` command which lists the resources deployed and their current statuses. The output of the command is also printed out as a part of App's status if enabled through `rawOptions` while creating the CR. 
+
+Though this command provides information about the resources created by the respective App CR, it does so by sending API requests during the reconciliation. Instead, using informers provides additional advantages of having real-time updates, efficient resource utilization and reduced load on API server.
+
+### Other Approaches:
+
+## Open Questions:
+
+1. Can using informers to watch resources increase cache size, potentially impacting the performance?
+2. Can the output in the `inspect` status field be combined with that of proposed `healthy` condition?
```

Comment:
I guess this is what needs to happen eventually. It doesn't make sense to have two status fields serve similar purpose. The `healthy` condition can just list all the resources (instead of just the failed/unhealthy) ones or vice-versa.

Before refactoring the proposal for the same, I would like to confirm the use case of `inspect` and make sure of the direction we would like to go.

Reply:
The `inspect` section initially just aggregated the statuses of all resources after a finished reconciliation. It was essentially the output of the `kapp inspect` command.
We disabled it by default in favour of reducing the number of API calls we make.

Since it is a separate feature altogether, I think we can work towards having a separate section for the additional information we want to surface.

Reply:
> Since it is a separate feature altogether, I think we can work towards having a separate section for the additional information we want to surface.

If we are watching all the resources in the cluster anyway, and triggering a reconcile - wondering if calling an `inspect` command on top of it is necessary. If so, this may also end up loading the API server - since I can expect more no. of reconciliations due to dependent resources.

Additionally, the second aspect in `inspect` and `healthy` showing conflicting information at any point in time to the user (I haven't looked into the codebase of `inspect`, but assuming controller client and the one used with kapp are different?). 
`Inspect` - would show a superset health status of all the resources.
`Healthy` - would show only the unhealthy resources. 

If we decide to support both of them, then probably we should make them exclusive?